File: ios/Sources/Wallet/DappProvider.swift
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

// The wallet acting as a CIP-0103 provider over the LAN — the iOS parity of
// the Android DappProviderController. A LanGrpcDappServer fronts a DappSession
// built from the wallet's own party and signing key.
//
// Additive: nothing in Send / Portfolio / Inbox changes. Scope is ping and
// sign-in — connect, listAccounts, getPrimaryAccount, status and signMessage
// are wired; prepareExecute is absent (it needs the JSON Ledger API pipeline),
// so a payment request comes back 4200 until that lands.

import CantonDappKit
import CantonDappLanKit
import CantonDappWalletKit
import CantonWalletKit
import Foundation
import SwiftUI

@MainActor
@Observable
final class DappProviderController {
    private(set) var listening = false
    private(set) var port: Int?
    private(set) var pending: PendingApproval?
    private(set) var lastActivity: String?

    private let account: DappWallet
    private let messageSigner: any DappMessageSigner
    private let networkId: String
    private var server: LanGrpcDappServer?

    init(account: DappWallet, messageSigner: any DappMessageSigner, networkId: String) {
        self.account = account
        self.messageSigner = messageSigner
        self.networkId = networkId
    }

    func start() async throws {
        guard !listening else { return }
        let session = DappSession(
            peer: DappPeer(id: "lan", name: "LAN dApp", verified: false),
            accounts: FixedAccounts(accounts: [account]),
            approver: Approver(controller: self),
            network: DappNetworkConfig(networkId: networkId),
            messageSigner: messageSigner
            // prepareExecute / ledgerApi intentionally absent for now.
        )
        // 0.0.0.0 so a dApp on this device reaches it at 127.0.0.1 and one on
        // another device over Wi-Fi.
        let server = LanGrpcDappServer(handler: session, host: "0.0.0.0")
        let boundPort = try await server.start(port: 0)
        self.server = server
        port = boundPort
        listening = true
        lastActivity = "Listening on port \(boundPort)"
    }

    func stop() {
        server?.shutdown()
        server = nil
        listening = false
        port = nil
        pending = nil
        lastActivity = "Stopped"
    }

    /// Called by the (Sendable) approver; raises the sheet and suspends the
    /// engine here until the user answers.
    fileprivate func requestApproval(_ request: DappApprovalRequest) async -> DappApproval {
        let (title, detail, accounts) = Self.describe(request, account: account)
        let decision = await withCheckedContinuation { (cont: CheckedContinuation<DappApproval, Never>) in
            self.pending = PendingApproval(title: title, detail: detail, accounts: accounts) { decision in
                cont.resume(returning: decision)
            }
        }
        pending = nil
        lastActivity = switch decision {
        case .approved: "Approved: \(title)"
        case .rejected: "Declined: \(title)"
        }
        return decision
    }

    private static func describe(
        _ request: DappApprovalRequest,
        account: DappWallet
    ) -> (String, String, [DappWallet]) {
        switch request {
        case .connection(let peer, let network, _):
            ("\(peer.name) wants to connect",
             "Share your account \(account.partyId.prefix(28))… on \(network.networkId)?",
             [account])
        case .message(let peer, _, let message):
            ("\(peer.name) wants a signature", message, [])
        case .transaction(let peer, let actAs, _, _):
            ("\(peer.name) wants to submit a transaction",
             "Acting as \(actAs.partyId.prefix(28))…", [])
        }
    }
}

/// One request awaiting the user. Completing it resumes the provider engine.
@MainActor
final class PendingApproval: Identifiable {
    let id = UUID()
    let title: String
    let detail: String
    private let accounts: [DappWallet]
    private let resolve: (DappApproval) -> Void

    init(title: String, detail: String, accounts: [DappWallet], resolve: @escaping (DappApproval) -> Void) {
        self.title = title
        self.detail = detail
        self.accounts = accounts
        self.resolve = resolve
    }

    func approve() { resolve(.approved(accounts: accounts)) }
    func reject() { resolve(.rejected(reason: "Declined in the wallet")) }
}

/// Sendable bridge from the provider engine to the MainActor controller.
private struct Approver: DappApprovalDelegate {
    let controller: DappProviderController

    func approve(_ request: DappApprovalRequest) async -> DappApproval {
        await controller.requestApproval(request)
    }
}

private struct FixedAccounts: DappAccountsSource {
    let accounts: [DappWallet]
    func accounts() async throws -> [DappWallet] { accounts }
}

// MARK: - UI

/// The wallet as a CIP-0103 provider over the LAN. Additive — it drives the
/// dApp side already shipped in the dApp reference app, and touches nothing in
/// Send / Portfolio / Inbox.
struct DappsView: View {
    @Environment(WalletModel.self) private var model

    var body: some View {
        NavigationStack {
            Form {
                if model.partyId == nil {
                    Text("Onboard the wallet first — a dApp connects to your party.")
                        .foregroundStyle(.secondary)
                } else {
                    Section {
                        Text("Let a CIP-0103 dApp connect to this wallet over the local "
                            + "network. The wallet approves every connection and signature; "
                            + "keys never leave it.")
                            .font(.callout)
                            .foregroundStyle(.secondary)
                    }
                    if let provider = model.dappProvider, provider.listening {
                        Section("Listening") {
                            LabeledContent("Port", value: "\(provider.port ?? 0)")
                            Text("A dApp on this device connects to 127.0.0.1:\(provider.port ?? 0); "
                                + "one on another device uses this phone's Wi-Fi address and the "
                                + "same port.")
                                .font(.footnote)
                                .foregroundStyle(.secondary)
                            Button("Stop listening", role: .destructive) { model.stopDappProvider() }
                        }
                        if let activity = provider.lastActivity {
                            Section { Text("Recent: \(activity)").font(.footnote) }
                        }
                    } else {
                        Section {
                            Button("Start listening") { model.startDappProvider() }
                        }
                    }
                }
            }
            .navigationTitle("dApps")
        }
        .sheet(item: Binding(
            get: { model.dappProvider?.pending },
            // Swiping the sheet away is a decline.
            set: { if $0 == nil { model.dappProvider?.pending?.reject() } }
        )) { pending in
            ApprovalSheet(pending: pending)
        }
    }
}

/// The connection / signature approval sheet — the wallet decides.
struct ApprovalSheet: View {
    let pending: PendingApproval

    var body: some View {
        VStack(spacing: 16) {
            Text(pending.title).font(.title2).bold()
            Text(pending.detail)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
            Divider()
            Button { pending.approve() } label: {
                Text("Approve").frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            Button(role: .destructive) { pending.reject() } label: {
                Text("Decline").frame(maxWidth: .infinity)
            }
            .buttonStyle(.bordered)
        }
        .padding(24)
        .presentationDetents([.medium])
    }
}

File: android/wallet-app/src/main/kotlin/io/github/vsima/canton/app/DappProvider.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import io.github.vsima.canton.dapp.DappWallet
import io.github.vsima.canton.dapp.wallet.DappApproval
import io.github.vsima.canton.dapp.wallet.DappApprovalDelegate
import io.github.vsima.canton.dapp.wallet.DappApprovalRequest
import io.github.vsima.canton.dapp.wallet.DappMessageSigner
import io.github.vsima.canton.dapp.wallet.DappNetworkConfig
import io.github.vsima.canton.dapp.wallet.DappPeer
import io.github.vsima.canton.dapp.wallet.DappSession
import io.github.vsima.canton.dapp.lan.LanGrpcDappServer
import java.net.InetSocketAddress
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * The wallet acting as a CIP-0103 **provider** over the LAN: a
 * [LanGrpcDappServer] fronting a [DappSession] built from the wallet's own
 * onboarded party and signing key.
 *
 * Additive by design — nothing in Send / Portfolio / Inbox changes. This is
 * the other half of the LAN transport whose dApp side already ships in the
 * dApp reference app: start listening here, and that app can connect.
 *
 * Scope for now: **ping and sign-in.** `connect`, `listAccounts`,
 * `getPrimaryAccount`, `status` and `signMessage` are wired; `prepareExecute`
 * is not (it needs the JSON Ledger API pipeline), so a payment request comes
 * back `4200` until that lands. That is the honest current state.
 */
class DappProviderController(
    private val account: DappWallet,
    private val messageSigner: DappMessageSigner,
    private val networkId: String,
    private val scope: CoroutineScope,
) {
    /** Whether the LAN provider is listening. */
    var listening by mutableStateOf(false)
        private set

    /** The port it is listening on, once started. */
    var port by mutableStateOf<Int?>(null)
        private set

    /** A request awaiting the user's decision, or null. Drives the sheet. */
    var pending by mutableStateOf<PendingApproval?>(null)
        private set

    /** A short line of the most recent activity, for the screen. */
    var lastActivity by mutableStateOf<String?>(null)
        private set

    private var server: LanGrpcDappServer? = null

    /**
     * One request the user must approve or reject. Completing the deferred
     * resumes the provider engine, which is suspended inside `approve`.
     */
    class PendingApproval(
        val title: String,
        val detail: String,
        private val decision: CompletableDeferred<DappApproval>,
        private val approvedAccounts: List<DappWallet>,
    ) {
        fun approve() {
            decision.complete(DappApproval.Approved(approvedAccounts))
        }

        fun reject() {
            decision.complete(DappApproval.Rejected("Declined in the wallet"))
        }
    }

    private val approver = DappApprovalDelegate { request ->
        val deferred = CompletableDeferred<DappApproval>()
        val (title, detail, accounts) = describe(request)
        // approve() runs on a gRPC thread; surface the sheet on the main
        // thread so Compose observes it, then suspend here until the user
        // answers. The engine is holding the request open in the meantime.
        withContext(Dispatchers.Main) {
            pending = PendingApproval(title, detail, deferred, accounts)
        }
        val decision = deferred.await()
        withContext(Dispatchers.Main) {
            pending = null
            lastActivity = when (decision) {
                is DappApproval.Approved -> "Approved: $title"
                is DappApproval.Rejected -> "Declined: $title"
            }
        }
        decision
    }

    private fun describe(request: DappApprovalRequest): Triple<String, String, List<DappWallet>> =
        when (request) {
            is DappApprovalRequest.Connection -> Triple(
                "${request.peer.name} wants to connect",
                "Share your account ${account.partyId.take(28)}… on ${request.network.networkId}?",
                listOf(account),
            )
            is DappApprovalRequest.Message -> Triple(
                "${request.peer.name} wants a signature",
                request.message,
                emptyList(),
            )
            is DappApprovalRequest.Transaction -> Triple(
                "${request.peer.name} wants to submit a transaction",
                "Acting as ${request.actAs.partyId.take(28)}…",
                emptyList(),
            )
        }

    /** Starts listening on all interfaces (loopback for a same-device dApp, LAN otherwise). */
    fun start() {
        if (listening) return
        val session = DappSession(
            peer = DappPeer(id = "lan", name = "LAN dApp", verified = false),
            accounts = { listOf(account) },
            approver = approver,
            network = DappNetworkConfig(networkId = networkId),
            messageSigner = messageSigner,
            // prepareExecute / ledgerApi intentionally absent for now — see the class note.
        )
        // 0.0.0.0 so a second device on the LAN can reach it too; a dApp on
        // this same device connects via 127.0.0.1.
        val started = LanGrpcDappServer(session, InetSocketAddress("0.0.0.0", 0)).start()
        server = started
        port = started.port
        listening = true
        lastActivity = "Listening on port ${started.port}"
    }

    fun stop() {
        server?.shutdown()
        server = null
        listening = false
        port = null
        pending = null
        lastActivity = "Stopped"
    }
}